### PR TITLE
Reintroduce support for the -processor flag for the translator

### PR DESCRIPTION
**Pull request by omerp**
Previous commit (bc5dbad0) removed jdt, along with support of the -processor flag.

Reimplement translator support for the flag, using the javac pre-processing step, along with relevant unit tests, as existing ones did not cover the -processor step (beyond it not failing).

### CODE REVIEW

**Comment by omerp** [comment]
Was added as an authorized contributor

File: translator/src/main/java/com/google/devtools/j2objc/javac/JavacParser.java [hunk 1]
```diff
@@ -185,16 +189,20 @@ public void parseFiles(Collection<String> paths, Handler handler, SourceVersion
...
 
   // Creates a javac environment from a collection of files and/or file objects.
-  private JavacEnvironment createEnvironment(List<File> files, List<JavaFileObject> fileObjects,
-      boolean processAnnotations) throws IOException {
+  private JavacEnvironment createEnvironment(
+      List<File> files,
+      List<JavaFileObject> fileObjects,
+      boolean processAnnotations,
+      String explicitProcessors)
+      throws IOException {
```
Comment (tomball) [inline translator/src/main/java/com/google/devtools/j2objc/javac/JavacParser.java]: Revert, processors are now picked up by getJavacOptions().

File: translator/src/main/java/com/google/devtools/j2objc/javac/JavacParser.java [hunk 2]
```diff
@@ -185,16 +189,20 @@ public void parseFiles(Collection<String> paths, Handler handler, SourceVersion
...
     JavaCompiler compiler = ToolProvider.getSystemJavaCompiler();
     DiagnosticCollector<JavaFileObject> diagnostics = new DiagnosticCollector<>();
     JavacFileManager fileManager = getFileManager(compiler, diagnostics);
-    List<String> javacOptions = getJavacOptions(processAnnotations);
+    List<String> javacOptions = getJavacOptions(processAnnotations, explicitProcessors);
```
Comment (tomball) [inline translator/src/main/java/com/google/devtools/j2objc/javac/JavacParser.java]: Revert.

File: translator/src/main/java/com/google/devtools/j2objc/javac/JavacParser.java [hunk 3]
```diff
@@ -235,13 +243,14 @@ public ProcessingResult processAnnotations(Iterable<String> fileArgs,
     PathClassLoader loader = new PathClassLoader(options.fileUtil().getClassPathEntries());
     loader.addPaths(options.getProcessorPathEntries());
     Iterator<Processor> serviceIterator = ServiceLoader.load(Processor.class, loader).iterator();
-    if (serviceIterator.hasNext()) {
+    String explicitProcessorList = options.getProcessors();
+    if (serviceIterator.hasNext() || explicitProcessorList != null) {
```
Comment (tomball) [inline translator/src/main/java/com/google/devtools/j2objc/javac/JavacParser.java]: Revert. Most annotation processing jars define a ServiceLoader entry in their manifest files, which is why this issue wasn't caught sooner. This test therefore needs to remain, and support for your use case is handled in getJavacOptions().

File: translator/src/main/java/com/google/devtools/j2objc/javac/JavacParser.java [hunk 4]
```diff
@@ -235,13 +243,14 @@ public ProcessingResult processAnnotations(Iterable<String> fileArgs,
...
       List<File> inputFiles = new ArrayList<>();
       for (ProcessingContext input : inputs) {
         inputFiles.add(new File(input.getFile().getAbsolutePath()));
       }
       try {
-        JavacEnvironment env = createEnvironment(inputFiles, null, true);
+        JavacEnvironment env = createEnvironment(inputFiles, null, true, explicitProcessorList);
```
Comment (tomball) [inline translator/src/main/java/com/google/devtools/j2objc/javac/JavacParser.java]: Revert.

File: translator/src/test/java/com/google/devtools/j2objc/J2ObjCTest.java [hunk 5]
```diff
@@ -223,17 +239,51 @@ public void testAnnotationProcessingWithProcessorPath() throws Exception {
...
+    options.fileUtil().getClassPathEntries().add(serviceProcessorPath);
+    options.fileUtil().getClassPathEntries().add(explicitProcessorPath);
+    options.setProcessors("com.google.devtools.j2objc.annotations.J2ObjCTestExplicitProcessor");
+
+    String examplePath = addSourceFile(EXAMPLE_JAVA_SOURCE, "annotations/Example.java");
+    J2ObjC.run(Collections.singletonList(examplePath), options);
+    assertErrorCount(0);
+
+    assertSpecifiedAnnotationProcessorOutput();
+    assertFalse("Overriden processor generated output", getTempFile("ProcessingResult.h").exists());
```
Comment (tomball) [inline translator/src/test/java/com/google/devtools/j2objc/J2ObjCTest.java]: Nit: "Overridden"

File: translator/src/main/java/com/google/devtools/j2objc/javac/JavacParser.java (hunk 3 above)
Comment (omerp) [reply translator/src/main/java/com/google/devtools/j2objc/javac/JavacParser.java]: @tomball - We have two types of processors here:
1) ServiceLoader processors (captured by serviceIterator.hasNext())
2) -processor processors

If we don't check for options.getProcessors() here and there are no service processors present, we'd never run the preprocessing step, even if we specify the -processor flag. This is why I'm currently checking both here and running pre-processing if either is present (which tests corroborate).

There is a slight change to the way j2objc currently operate, as the -processor flag is currently ignored and will now override any ServiceLoader processor, due to the interface defined by javac (and also documented in the j2objc -help of the -processor flag)

**Comment by omerp** [comment]
Updated the commit. If by the license click-through you mean the CLA - it should be alright now

**Pull request merged**